sw/device/lib/dif/dif_aes.c

```diff
@@ -267,10 +286,61 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
     return kDifError;
   }
 
-  for (int i = 0; i < AES_DATA_OUT_MULTIREG_COUNT; ++i) {
-    ptrdiff_t offset = AES_DATA_OUT_0_REG_OFFSET + (i * sizeof(uint32_t));
+  aes_read_multireg(aes, data->data, AES_DATA_OUT_MULTIREG_COUNT,
+                    AES_DATA_OUT_0_REG_OFFSET);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_process_data(const dif_aes_t *aes,
+                                  const dif_aes_data_t *plain_text,
+                                  dif_aes_data_t *cipher_text,
+                                  size_t block_count) {
+  if (aes == NULL || plain_text == NULL || cipher_text == NULL ||
+      block_count == 0) {
+    return kDifBadArg;
+  }
+
+  // The algorithm below just makes sense for at least 2 blocks. Otherwise
+  // it is better to use the `load_data` and `read_output` functions.
+  if (block_count < 2) {
+    DIF_RETURN_IF_ERROR(dif_aes_load_data(aes, plain_text[0]));
+    return dif_aes_read_output(aes, cipher_text);
+  }
+
+  // Ensure that the INPUT_READY bit in STATUS is 1.
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  // Write Input Data Block 0 to the Input Data registers DATA_IN_0 - DATA_IN_3.
+  aes_set_multireg(aes, plain_text[0].data, AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  // Wait for the INPUT_READY bit in STATUS to become 1, i.e. wait for the AES
+  // unit to load Input Data Block 0 into the internal state register and start
+  // operation.
+  AES_WAIT_FOR_STATUS(aes, AES_STATUS_INPUT_READY_BIT, true);
+  // Then for every Data Block I=0,..,N-3, software must:
+  for (size_t i = 0, j = 1; i < block_count; ++i, ++j) {
```

Comment:
Why have two induction variables, when you can use `i + 1` throughout? Two induction variables can become harder to maintain if they do not remain in lockstep like this.